snuba/utils/streams/abstract.py

```diff
@@ -0,0 +1,137 @@
+    to their implementation.
+    """
+
+
+class EndOfStream(ConsumerError, Generic[TStream]):
+    """
+    Raised when there are no more messages to consume from the stream.
+    """
+
+    def __init__(self, stream: TStream):
+        self.stream = stream
+
+
+class Consumer(ABC, Generic[TStream, TOffset, TValue]):
+    """
+    An abstract class that provides methods to consume messages from a
+    multiplexed collection of streams.
+
+    This interface is heavily "inspired" by the Confluent Kafka Consumer
+    implementation, but only exposes a limited set of the available methods
+    and several method signatures and/or return values differ -- these
+    changes are intended to improve the ease of use of the consumer, as well
+    as make development and testing more straightforward. Most code that uses
+    the Confluent driver can be modified to use this interface with minimal
+    changes, but it is not intended to be a drop-in replacement for the
+    Confluent implementation.
+
+    All methods are blocking unless otherise noted in the documentation for
+    that method.
+    """
+
+    @abstractmethod
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TStream]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[TStream]], None]] = None,
+    ) -> None:
+        """
+        Subscribe to topic streams. This replaces a previous subscription.
+        This method does not block. The subscription may not be fulfilled
+        immediately: instead, the ``on_assign`` and ``on_revoke`` callbacks
+        are called when the subscription state changes with the updated
+        assignment for this consumer.
+
+        Raises a ``RuntimeError`` if called on a closed consumer.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def poll(
+        self, timeout: Optional[float] = None
+    ) -> Optional[Message[TStream, TOffset, TValue]]:
+        """
+        Return the next message available to be consumed, if one is
+        available. If no message is available, this method will block up to
+        the ``timeout`` value before returning ``None``. A timeout of
+        ``0.0`` represents "do not block", while a timeout of ``None``
+        represents "block until a message is available (or forever)".
+
+        Calling this method may also invoke subscription state change
+        callbacks.
+
+        If the backend supports it, this method may raise a ``EndOfStream``
+        error if there are no messages available to be consumed (and the
+        backend is configured to do so.) This may also raise
+        ``ConsumerError`` or one of it's subclasses, the meaning of which are
+        implementation-dependent.
+
+        Raises a ``RuntimeError`` if called on a closed consumer.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def commit(self) -> Mapping[TStream, TOffset]:
```

Comment:
We will have to be able to commit specific messages or specific offsets in the implementation (strict consumer). Is this going to be supported ? Also here we have a good opportunity to do better than the confluent implementation by having a dedicated commit_message and commit_offset method

Reply:
I'm intentionally not considering the strict consumer as part of this change, so it may or may not be supported. I don't think that's a decision that needs to be made right now.

My guess is we'll probably need to handle something like that in the future but it's not part of the batching use cases (at least at the moment.)

Reply:
Also, to be clear, I think we can consolidate all of the consumer access through this class eventually, I just want to solve one problem at a time right now instead of adding more responsibility here which will make this change even more complicated and risky.